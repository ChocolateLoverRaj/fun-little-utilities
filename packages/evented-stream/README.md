# Evented Stream

Collection of evented-stream packages

## License

[MIT](https://git.io/fxKXN) © [Hoàng Văn Khải](https://github.com/KSXGitHub)
