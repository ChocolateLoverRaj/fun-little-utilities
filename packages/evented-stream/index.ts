import * as types from 'evented-stream-types'
import iterateEventedStream from 'iterate-evented-stream'
import combineEventedStream from 'combine-evented-stream'

export * from 'evented-stream-types'
export { iterateEventedStream, combineEventedStream, types }

export const iterate = iterateEventedStream
export const combine = combineEventedStream
