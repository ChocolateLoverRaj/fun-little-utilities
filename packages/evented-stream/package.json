{
  "name": "evented-stream",
  "version": "0.1.0",
  "description": "Collection of evented-stream packages",
  "author": "Hoàng Văn Khải <hvksmr1996@gmail.com>",
  "license": "MIT",
  "homepage": "https://ksxnodeapps.github.io/fun-little-utilities/",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/ksxnodeapps/fun-little-utilities.git"
  },
  "bugs": {
    "url": "https://github.com/ksxnodeapps/fun-little-utilities/issues"
  },
  "keywords": [
    "stream",
    "iterate",
    "asynchronous",
    "event",
    "async",
    "iterable"
  ],
  "dependencies": {
    "tslib": "^1.10.0",
    "iterate-evented-stream": "^0.1.0",
    "combine-evented-stream": "^0.1.0",
    "evented-stream-types": "^0.1.0",
    "@types/node": "^12.7.8"
  },
  "engines": {
    "node": ">= 8.9.0"
  }
}
