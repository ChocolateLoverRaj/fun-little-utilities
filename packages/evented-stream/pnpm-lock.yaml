dependencies:
  '@types/node': 13.11.0
  combine-evented-stream: 'link:../combine-evented-stream'
  evented-stream-types: 'link:../evented-stream-types'
  iterate-evented-stream: 'link:../iterate-evented-stream'
  tslib: 1.11.1
lockfileVersion: 5.1
packages:
  /@types/node/13.11.0:
    dev: false
    resolution:
      integrity: sha512-uM4mnmsIIPK/yeO+42F2RQhGUIs39K2RFmugcJANppXe6J1nvH87PvzPZYpza7Xhhs8Yn9yIAVdLZ84z61+0xQ==
  /tslib/1.11.1:
    dev: false
    resolution:
      integrity: sha512-aZW88SY8kQbU7gpV19lN24LtXh/yD4ZZg6qieAJDDg+YBsJcSmLGK9QpnUjAKVG/xefmvJGd1WUmfpT/g6AJGA==
specifiers:
  '@types/node': ^13.11.0
  combine-evented-stream: ^0.1.7
  evented-stream-types: ^0.1.7
  iterate-evented-stream: ^0.1.8
  tslib: ^1.11.1
