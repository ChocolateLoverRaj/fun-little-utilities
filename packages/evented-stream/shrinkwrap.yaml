dependencies:
  '@types/node': 10.12.18
  combine-evented-stream: 'link:../combine-evented-stream'
  evented-stream-types: 'link:../evented-stream-types'
  iterate-evented-stream: 'link:../iterate-evented-stream'
  tslib: 1.9.3
packages:
  /@types/node/10.12.18:
    dev: false
    resolution:
      integrity: sha512-fh+pAqt4xRzPfqA6eh3Z2y6fyZavRIumvjhaCL753+TVkGKGhpPeyrJG2JftD0T9q4GF00KjefsQ+PQNDdWQaQ==
  /tslib/1.9.3:
    dev: false
    resolution:
      integrity: sha512-4krF8scpejhaOgqzBEcGM7yDIEfi0/8+8zDRZhNZZ2kjmHJ4hv3zCbQWxoJGz1iw5U0Jl0nma13xzHXcncMavQ==
registry: 'https://registry.npmjs.org/'
shrinkwrapMinorVersion: 9
shrinkwrapVersion: 3
specifiers:
  '@types/node': ^10.12.18
  combine-evented-stream: ^0.0.0
  evented-stream-types: ^0.0.0
  iterate-evented-stream: ^0.0.1
  tslib: ^1.9.3
