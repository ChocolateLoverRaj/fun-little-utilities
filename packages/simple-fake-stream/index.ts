export * from './method-call'
export * from './stream'
export * from './string'
