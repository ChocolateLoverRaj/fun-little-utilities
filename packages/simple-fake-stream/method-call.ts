// tslint:disable:no-unnecessary-qualifier

import { EventedStream } from 'evented-stream-types'
import { AnyClass } from './utils'

export enum MethodName {
  addListener = 'addListener',
  removeListener = 'removeListener',
  on = 'on',
  once = 'once',
  write = 'write',
}

export namespace MethodName {
  export type EventModifier =
    | MethodName.addListener
    | MethodName.removeListener
    | MethodName.on
    | MethodName.once
}

export type MethodCall<Chunk, Err> =
  | MethodCall.EventModifier<Chunk, Err>
  | MethodCall.Write<Chunk>

export namespace MethodCall {
  export interface Base {
    readonly name: MethodName
  }

  export type EventModifier<Chunk, Err> =
    | EventModifier.Data<Chunk>
    | EventModifier.Error<Err>
    | EventModifier.Close

  export namespace EventModifier {
    export interface Base extends MethodCall.Base {
      readonly name: MethodName.EventModifier
      readonly event: string
      readonly listener: Function
    }

    export interface Data<Chunk> extends Base {
      readonly event: 'data'
      readonly listener: EventedStream.DataEventListener<Chunk>
    }

    export interface Error<Err> extends Base {
      readonly event: 'error'
      readonly listener: EventedStream.ErrorEventListener<Err>
    }

    export interface Close extends Base {
      readonly event: 'close'
      readonly listener: EventedStream.CloseEventListener
    }
  }

  export interface Write<Chunk> extends Base {
    readonly name: MethodName.write
    readonly data: Chunk
  }
}

export type MethodCallInstance<Chunk, Err> =
  | MethodCallInstance.EventModifier<Chunk, Err>
  | MethodCallInstance.Write<Chunk>

export namespace MethodCallInstance {
  function rename<Class extends AnyClass>(Base: Class): Class {
    return class MethodCallInstance extends Base {}
  }

  export abstract class Base<Name extends MethodName = MethodName> implements MethodCall.Base {
    constructor(
      public readonly name: Name,
    ) {}
  }

  export type EventModifier<Chunk, Err> =
    | EventModifier.Data<Chunk>
    | EventModifier.Error<Err>
    | EventModifier.Close

  export namespace EventModifier {
    function bindEvent<Event extends string>(event: Event) {
      return class<Listener extends Function> extends Base<Event, Listener> {
        constructor(name: MethodName.EventModifier, listener: Listener) {
          super(name, event, listener)
        }
      }
    }

    export abstract class Base<Event extends string, Listener extends Function>
      extends MethodCallInstance.Base<MethodName.EventModifier>
      implements MethodCall.EventModifier.Base {
      constructor(
        name: MethodName.EventModifier,
        public readonly event: Event,
        public readonly listener: Listener,
      ) {
        super(name)
      }
    }

    @rename
    export class Data<Chunk> extends bindEvent('data')<EventedStream.DataEventListener<Chunk>>
      implements MethodCall.EventModifier.Data<Chunk> {}

    @rename
    export class Error<Err> extends bindEvent('error')<EventedStream.ErrorEventListener<Err>>
      implements MethodCall.EventModifier.Error<Err> {}

    @rename
    export class Close extends bindEvent('close')<EventedStream.CloseEventListener>
      implements MethodCall.EventModifier.Close {}
  }

  @rename
  export class Write<Chunk> extends Base<MethodName.write> implements MethodCall.Write<Chunk> {
    constructor(
      public readonly data: Chunk,
    ) {
      super(MethodName.write)
    }
  }
}
