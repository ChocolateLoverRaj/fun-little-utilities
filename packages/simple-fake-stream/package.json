{
  "name": "simple-fake-stream",
  "version": "0.1.5",
  "description": "Create fake streams for testing purpose",
  "author": "Hoàng Văn Khải <hvksmr1996@gmail.com>",
  "license": "MIT",
  "homepage": "https://ksxnodeapps.github.io/fun-little-utilities/",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/ksxnodeapps/fun-little-utilities.git"
  },
  "bugs": {
    "url": "https://github.com/ksxnodeapps/fun-little-utilities/issues"
  },
  "keywords": [
    "stream",
    "readable",
    "writable",
    "fake",
    "test"
  ],
  "dependencies": {
    "monorepo-shared-assets": "^0.0.19",
    "tslib": "^1.10.0",
    "evented-stream-types": "^0.1.5",
    "@types/node": "^13.5.2"
  },
  "engines": {
    "node": ">= 8.9.0"
  }
}
