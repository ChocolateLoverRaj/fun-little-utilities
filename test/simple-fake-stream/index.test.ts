import * as assets from 'monorepo-shared-assets'
import * as evtstr from 'evented-stream'
import * as sfs from 'simple-fake-stream'
import getAsyncArray = assets.asyncIter.fns.getArray

abstract class InitBase {
  public readonly stream = new sfs.StreamInstance<string>()

  constructor() {
    this.init(this.stream)
  }

  protected abstract init(stream: InitBase['stream']): void
}

it('MethodName matches snapshot', () => {
  expect(sfs.MethodName).toMatchSnapshot()
})

describe('without method calls', () => {
  class Init extends InitBase {
    protected init() {
      return undefined
    }
  }

  it('StreamDatabase::getMethodCalls() is empty', () => {
    expect(new Init().stream.getMethodCalls()).toEqual([])
  })

  it('StreamDatabase::getChunks() is empty', () => {
    expect(new Init().stream.getChunks()).toEqual([])
  })
})

describe('when WritableStream::write is called', () => {
  class Init extends InitBase {
    protected init(stream: Init['stream']): void {
      stream.write('foo')
      stream.write('bar')
      stream.write('baz')
    }
  }

  it('StreamDatabase::getMethodCalls() matches snapshot', () => {
    expect(new Init().stream.getMethodCalls()).toMatchSnapshot()
  })

  it('StreamDatabase::getChunks() gives expected result', () => {
    expect(new Init().stream.getChunks()).toEqual(['foo', 'bar', 'baz'])
  })

  it('stream is iterable', async () => {
    const { stream } = new Init()
    expect(await getAsyncArray(stream)).toEqual(stream.getChunks())
  })

  it('getString() gives expected result', () => {
    expect(sfs.getString(new Init().stream)).toBe('foobarbaz')
  })

  it('works well with iterate-evented-stream', async () => {
    const { stream } = new Init()
    const iterator = evtstr.iterate(stream)
    void stream.asyncEmit('close')
    const array = await getAsyncArray(iterator)
    expect(array).toEqual(stream.getChunks())
  })
})

describe('when EventedReadableStream::{addListener,removeListener,on,once} is called', () => {
  class FnSetByMethod {
    public addListener(): void {
      return undefined
    }
    public removeListener(): void {
      return undefined
    }
    public on(): void {
      return undefined
    }
    public once(): void {
      return undefined
    }
  }

  class FnSetByEvent {
    public readonly data = new FnSetByMethod()
    public readonly error = new FnSetByMethod()
    public readonly close = new FnSetByMethod()
  }

  const listeners = new FnSetByEvent()

  class Init extends InitBase {
    protected init(stream: Init['stream']): void {
      stream.addListener('data', listeners.data.addListener)
      stream.removeListener('data', listeners.data.removeListener)
      stream.on('data', listeners.data.on)
      stream.once('data', listeners.data.once)

      stream.addListener('error', listeners.error.addListener)
      stream.removeListener('error', listeners.error.removeListener)
      stream.on('error', listeners.error.on)
      stream.once('error', listeners.error.once)

      stream.addListener('close', listeners.close.addListener)
      stream.removeListener('close', listeners.close.removeListener)
      stream.on('close', listeners.close.on)
      stream.once('close', listeners.close.once)
    }
  }

  it('StreamDatabase::getMethodCalls() gives expected result', () => {
    const { addListener, removeListener, on, once } = sfs.MethodName
    const { Data, Error, Close } = sfs.MethodCallInstance.EventModifier

    expect(new Init().stream.getMethodCalls()).toEqual([
      new Data(addListener, listeners.data.addListener),
      new Data(removeListener, listeners.data.removeListener),
      new Data(on, listeners.data.on),
      new Data(once, listeners.data.once),

      new Error(addListener, listeners.error.addListener),
      new Error(removeListener, listeners.error.removeListener),
      new Error(on, listeners.error.on),
      new Error(once, listeners.error.once),

      new Close(addListener, listeners.close.addListener),
      new Close(removeListener, listeners.close.removeListener),
      new Close(on, listeners.close.on),
      new Close(once, listeners.close.once),
    ])
  })
})
