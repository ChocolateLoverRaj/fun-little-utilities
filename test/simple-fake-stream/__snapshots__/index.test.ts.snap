// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MethodName matches snapshot 1`] = `
Object {
  "addListener": "addListener",
  "on": "on",
  "once": "once",
  "removeListener": "removeListener",
  "write": "write",
}
`;

exports[`when WritableStream::write is called StreamDatabase::getMethodCalls() matches snapshot 1`] = `
Array [
  MethodCallInstance {
    "data": "foo",
    "name": "write",
  },
  MethodCallInstance {
    "data": "bar",
    "name": "write",
  },
  MethodCallInstance {
    "data": "baz",
    "name": "write",
  },
]
`;
